bootstrap/kubeadm/api/v1beta1/kubeadm_types.go

```diff
@@ -512,6 +512,143 @@ type BootstrapTokenDiscovery struct {
+	// +optional
+	KubeConfig *FileDiscoveryKubeConfig `json:"kubeConfig,omitempty"`
+}
+
+// FileDiscoveryKubeConfig contains elements describing how to generate the kubeconfig for bootstrapping.
+type FileDiscoveryKubeConfig struct {
+	// Cluster contains information about how to communicate with the kubernetes cluster.
+	//
+	// By default the following fields are automatically populated:
+	// - Name with the Cluster's name.
+	// - Server with the Cluster's ControlPlaneEndpoint.
+	// - CertificateAuthorityData with the Cluster's CA certificate.
+	// +optional
+	Cluster *KubeConfigCluster `json:"cluster,omitempty"`
+
+	// User contains information that describes identity information.
+	// This is use to tell the kubernetes cluster who you are.
+	User KubeConfigUser `json:"user"`
+}
+
+// KubeConfigCluster contains information about how to communicate with a kubernetes cluster.
+//
+// Adapted from clientcmdv1.Cluster.
+type KubeConfigCluster struct {
+	// Server is the address of the kubernetes cluster (https://hostname:port).
+	Server string `json:"server"`
+	// TLSServerName is used to check server certificate. If TLSServerName is empty, the hostname used to contact the server is used.
+	// +optional
+	TLSServerName string `json:"tls-server-name,omitempty"`
+	// InsecureSkipTLSVerify skips the validity check for the server's certificate. This will make your HTTPS connections insecure.
+	// +optional
+	InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
+	// CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+	// +optional
+	CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
+	// ProxyURL is the URL to the proxy to be used for all requests made by this
+	// client. URLs with "http", "https", and "socks5" schemes are supported.  If
+	// this configuration is not provided or the empty string, the client
+	// attempts to construct a proxy configuration from http_proxy and
+	// https_proxy environment variables. If these environment variables are not
+	// set, the client does not attempt to proxy requests.
+	//
+	// socks5 proxying does not currently support spdy streaming endpoints (exec,
+	// attach, port forward).
+	// +optional
+	ProxyURL string `json:"proxy-url,omitempty"`
+}
+
+// KubeConfigUser contains information that describes identity information.
+// This is use to tell the kubernetes cluster who you are.
+//
+// Adapted from clientcmdv1.AuthInfo.
+type KubeConfigUser struct {
+	// ClientCertificateData contains PEM-encoded data from a client cert file for TLS. Overrides ClientCertificate
+	// +optional
+	ClientCertificateData []byte `json:"client-certificate-data,omitempty"`
+	// ClientKeyData contains PEM-encoded data from a client key file for TLS. Overrides ClientKey
+	// +optional
+	ClientKeyData []byte `json:"client-key-data,omitempty" datapolicy:"security-key"`
```

Comment:
Are we sure we want to leak this in a KubeadmConfig / KubeadmConfigTemplate?

Same security issue for the Token field below

Reply:
Sort of related, how does this work with the existing Discovery.BootstrapToken? Are we actually writing this to the KubeadmConfig / apiserver / etcd. Or is it there just in memory and then used to calculate the bootstrap secret?

Reply:
nope, having the key here is not secure.

one concern that i have.
the correct way to do file discovery with kubeadm is to feed unique cert/key from a kubeconfig on disk.
by embedding such a file as part of CAPI's kubeadmconfigspec.joinconfiguration it seems like CAPI users will use the same kubeconfig to join multiple nodes, which i don't think we can claim as a best practice.

Reply:
Remove `ClientCertificateData` `ClientKeyData` and `Token`?

Reply:
+1

Independent of that, would be good to know if we defer patch the bootstrap token today onto a KubeadmConfig. At a first glance the code looks like it, but I didn't see it in CI. I'll probably give it a try in tilt

Reply:
All done

Reply:
Hm okay was looking at the wrong place in testgrid before.

https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1808104956754923520/artifacts/clusters/bootstrap/resources/quick-start-rnx1ap/KubeadmConfig/quick-start-z3fd7b-md-0-skckq-jmh8x-zx4mw.yaml

Not great?

Reply:
One of the main reasons to switch to bootstrap file is exactly this

Reply:
I would say there is no need for us to actuallly write the token to the apiserver, right? :) (we can also only use it to write the bootstrap secret)

Reply:
(Maybe we should discuss this in a separate issue?)